src/v/iceberg/transaction.cc

```diff
@@ -0,0 +1,69 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/transaction.h"
+
+#include "iceberg/schema.h"
+#include "iceberg/table_requirement.h"
+#include "iceberg/table_update_applier.h"
+#include "iceberg/update_schema_action.h"
+
+namespace iceberg {
+
+ss::future<transaction::txn_outcome>
+transaction::apply(std::unique_ptr<action> a) {
+    if (error_.has_value()) {
+        co_return error_.value();
+    }
+    auto u = co_await std::move(*a).build_updates();
```

Comment:
why do we move out of `a` here? Is it to trigger the destructor of `*a`?

Reply:
I see from the `build_updates` signature it is supposed to be called on an rvalue so it can only be called once?

Reply:
Yes, exactly. I tried to make the interface strongly opinionated to avoid running actions multiple times accidentally